src/server/hset_family.cc

```diff
@@ -1271,10 +1271,12 @@ void HSetFamily::HRandField(CmdArgList args, const CommandContext& cmd_cntx) {
   auto* rb = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
   OpResult<StringVec> result = cmd_cntx.tx->ScheduleSingleHopT(std::move(cb));
   if (result) {
-    if ((result->size() == 1) && (args.size() == 1))
+    if (result->size() == 1 && args.size() == 1)
       rb->SendBulkString(result->front());
+    else if (with_values)
+      rb->SendBulkStrArrAsPairs(*result);
```

Comment:
I would make this a lambda and call it here instead of moving it within reply_builder. My rationale for this is that I suspect, that redis/valkey itself has a bug here. From the docs:

```
RESP2 Reply
[Array reply](https://redis.io/docs/latest/develop/reference/protocol-spec#arrays): a list of fields and their values when count and WITHVALUES were both used.

and 
RESP3 Reply
selected field when the count option is not used
[Array reply](https://redis.io/docs/latest/develop/reference/protocol-spec#arrays): a list of fields and their values when count and WITHVALUES were both used.
```

Reply:
Maybe something like:

```
if resp 2 -> rb->SendBulkString()
else {
  DCHECK(strs_size % 2 == 0) << "unexpected size of strings " << strs_size << ", expected pairs";
  StartArray(strs_size / 2);
  for (size_t i = 0; i < strs_size; i += 2) {
    StartArray(2);
    SendBulkString(strs[i]);
    SendBulkString(strs[i + 1]);
  }
}
```

Reply:
changed. I had to copy the reply scope object to the scope to ensure the message gets flushed.